### PR TITLE
Add report buffer and |futureOnly| option.

### PR DESCRIPTION
**Pull request by paulmeyer90**

### CODE REVIEW

File: index.src.html [hunk 1]
```diff
@@ -757,8 +757,22 @@ <h3 id="try-delivery" algorithm>
...
+  Each <a>environment settings object</a> has a <dfn>registered reporting
+  observer list</dfn>, which is a list of <a>reporting observers</a>.
+
+  To <dfn>register</dfn> a <a>reporting observer</a> means to add it to
+  the <a>registered reporting observer list</a>.
+
+  Each <a>environment settings object</a> has a <dfn>report buffer</dfn>, which
+  is an ordered list of <a>reports</a> that have been generated in
+  that <a>environment settings object</a>. This list is initially empty, and the
+  reports are stored in the same chronological order in which they are
```
Comment (RByers) [inline index.src.html]: nit: "same order" is sufficient (chronological implies something about a clock to me).

File: index.src.html [hunk 2]
```diff
@@ -796,15 +809,17 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
...
   Each {{ReportingObserver}} object has these associated concepts:
-    - A <dfn for=ReportingObserver>callback</dfn> set on creation.
-    - An {{ReportingObserverOptions}} dictionary called
+    - A <dfn for=ReportingObserver>callback</dfn> function set on creation.
+    - A {{ReportingObserverOptions}} dictionary called
       <dfn for=ReportingObserver>options</dfn>.
     - A list of {{Report}} objects called the <dfn for=ReportingObserver>report
-      list</dfn> that is initially empty.
+      list</dfn>, which is initially empty if the <a>context object</a> has a
+      {{futureOnly}} <a>option</a> that is true, or a copy of the <a>report
```
Comment (RByers) [inline index.src.html]: I think the preferred style is to describe non-trivial initialization like this in terms of an algorithm.  I.e. think of this section as your variable declarations.  It doesn't make sense to say "initially a copy of the report buffer .. if the context object has a futureOnly option" because *initially* the ReportingObserver has no options.  I think you want to instead update the algorithm for the `observe` method to say something like "if the futureOnly option is false, set the report list to a copy of the report buffer" etc.  It's probably important to define exactly when this happens because it may be observable (which reports are present).

File: index.src.html [hunk 3]
```diff
@@ -783,11 +797,10 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
 callback ReportingObserverCallback = void (sequence&lt;Report> reports, ReportingObserver observer);
 
 dictionary ReportingObserverOptions {
-  sequence&lt;DOMString> types;
+    sequence&lt;DOMString> types;
+    boolean futureOnly;
```
Comment (RByers) [inline index.src.html]: `futureOnly` being not present and being specified as false should behave identically, right?  You can make this clear and simplify some spec text by specifying a default value (`= false`).  See [here](https://heycam.github.io/webidl/#dfn-dictionary).

File: index.src.html [hunk 4]
```diff
@@ -816,27 +831,29 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
   
   1. Set the <a>context object</a>'s <a>options</a> to |options|.
   
-  2. If it is not already present, add the <a>context object</a> to the 
-     list of <a>registered reporting observers</a> for the <a>environment
-     settings object</a> associated with the <a>context object</a>.
+  2. If it is not already registered, <a>register</a> the <a>context object</a>
+     with its associated <a>environment settings object</a>.
```
Comment (RByers) [inline index.src.html]: why the asymmetry between "register" having a definition, but removing from the list being defined inline (i.e. why use a helper function for one but not the other)?  It seems simpler to me to stick with the original definition (no helper function) but that might be only because I haven't gotten to the part yet where this is helpful :-)

File: index.src.html [hunk 5]
```diff
@@ -816,27 +831,29 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
   
   1. Set the <a>context object</a>'s <a>options</a> to |options|.
   
-  2. If it is not already present, add the <a>context object</a> to the 
-     list of <a>registered reporting observers</a> for the <a>environment
-     settings object</a> associated with the <a>context object</a>.
+  2. If it is not already registered, <a>register</a> the <a>context object</a>
```
Comment (RByers) [inline index.src.html]: The concept "registered" is not, I think, precisely defined.  I.e. you've written a function for "register" but not 'isRegistered".

File: index.src.html [hunk 6]
```diff
@@ -816,27 +831,29 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
...
-  1. If present, remove the <a>context object</a> from the list of
-     <a>registered reporting observers</a> for the <a>environment settings
-     object</a> associated with the <a>context object</a>.
+
+  1. If the <a>context object</a> is not registered, stop.
+
+  2. Remove the <a>context object</a> from the <a>registered reporting observer
+     list</a> of its associated <a>environment settings object</a>.
+
+  3. Execute the <a href="#invoke-observers">invoke reporting observers</a>
```
Comment (RByers) [inline index.src.html]: Why is this necessary?  In most cases a task is already queued when the report list is updated.

Is it just for the case where `futureOnly` was false and no other reports have been queued yet?  I don't think we want to delay callback invocation those reports until `disconnect`.  Instead `observe` should probably be queuing a task like this when the buffer starts with some entries.

File: index.src.html [hunk 7]
```diff
@@ -757,8 +757,22 @@ <h3 id="try-delivery" algorithm>
...
+  object.
+
+  Each <a>environment settings object</a> has a <dfn>registered reporting
+  observer list</dfn>, which is a list of <a>reporting observers</a>.
+
+  To <dfn>register</dfn> a <a>reporting observer</a> means to add it to
+  the <a>registered reporting observer list</a>.
+
+  Each <a>environment settings object</a> has a <dfn>report buffer</dfn>, which
+  is an ordered list of <a>reports</a> that have been generated in
```
Comment (RByers) [inline index.src.html]: You don't say where things get added to the report buffer, or how overflow is handled.  I think that should just be a matter of adding a step to the 'notify' algorithm.  I think other observer specs have a pattern you can follow?

**Review by RByers — CHANGES_REQUESTED** [review]
This looks like a great start, thanks!

File: index.src.html (hunk 5 above)
Comment (paulmeyer90) [reply index.src.html]: Okay, I will make the definition of "registered" explicit and precise.

File: index.src.html (hunk 6 above)
Comment (paulmeyer90) [reply index.src.html]: I think you're right, and this shouldn't be necessary. I will remove step 3 here and ensure a task is queued when buffered reports are added to new observers.

File: index.src.html (hunk 1 above)
Comment (paulmeyer90) [reply index.src.html]: Done.

File: index.src.html (hunk 3 above)
Comment (paulmeyer90) [reply index.src.html]: Okay, done.

File: index.src.html (hunk 2 above)
Comment (paulmeyer90) [reply index.src.html]: I have added this to the algorithm for <code>observe</code>, and have also now rectified in that algorithm the fact that <code>report buffer</code> contains <code>report</code>s, but <code>report list</code> contains <code>Report</code> objects.

File: index.src.html (hunk 4 above)
Comment (paulmeyer90) [reply index.src.html]: I just thought that using the "registered" term made the many references to this property simpler and more clear, e.g. a step can be worded "If registered" instead of "If in the registered reporting observers list of the associated environment settings object". I then created the "register" helper to go with it, but I guess I'm only using it in one place, so I can remove it to fix the asymmetry.

File: index.src.html (hunk 7 above)
Comment (paulmeyer90) [reply index.src.html]: I have added to the notify algorithm how the reports get added to the report buffer, and specified how overflow is handled.

File: index.src.html [hunk 8]
```diff
@@ -757,9 +757,22 @@ <h3 id="try-delivery" algorithm>
...
-  Each <a>environment settings object</a> has an associated list of
-  <dfn>registered reporting observers</dfn> ({{ReportingObserver}} objects).
-   
+
+  A <dfn>reporting observer</dfn> observes some types of <a>reports</a> from
+  JavaScript, and is represented in JavaScript by the {{ReportingObserver}}
+  object.
+
+  Each <a>environment settings object</a> has a <dfn>registered reporting
+  observer list</dfn>, which is a list of <a>reporting observers</a>.
```
Comment (clelland) [inline index.src.html]: I think what you actually want is an ordered set (https://infra.spec.whatwg.org/#ordered-set) since the observe() method checks for duplicates before adding to this list. (You can still call it a 'list', I think, since an ordered set *is a* list)

File: index.src.html [hunk 9]
```diff
@@ -769,103 +782,123 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
...
+    - A {{ReportingObserverOptions}} dictionary called
       <dfn for=ReportingObserver>options</dfn>.
     - A list of {{Report}} objects called the <dfn for=ReportingObserver>report
-      list</dfn> that is initially empty.
+      list</dfn>, which is initially empty.
 
   The <dfn constructor for=ReportingObserver><code>
   ReportingObserver(|callback|)</code></dfn> constructor, when invoked, must
   create a new {{ReportingObserver}} object with <a>callback</a> set to
   |callback|, and then return it.
```
Comment (clelland) [inline index.src.html]: Should these constructor steps say what `options` is initialized to?

File: index.src.html [hunk 10]
```diff
@@ -769,103 +782,123 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
...
-  
+
+  2. Let |settings| be the <a>environment settings object</a> associated with
+     the <a>context object</a>
+
+  3. If it is not already registered, add the <a>context object</a> to
+     the <a>registered reporting observer list</a> of its
+     associated <a>environment settings object</a>.
+
+  4. If the <a>context object's</a> {{futureOnly}} <a>option</a> is true, stop.
```
Comment (clelland) [inline index.src.html]: Can  you use "return" instead of "stop"?

File: index.src.html [hunk 11]
```diff
@@ -769,103 +782,123 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
...
+
+  5. For each <a>report</a> |report| in the <a>report buffer</a> associated with
```
Comment (clelland) [inline index.src.html]: I think you can remove the `<a>report</a>` here. The pattern is just

> For each <var>item</var> in <var>sequence</var>, ...

File: index.src.html [hunk 12]
```diff
@@ -769,103 +782,123 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
...
+     |settings|, execute [[#add-report]] with |report| and the <a>context object</a>.
+
+  6. <a>Queue a task</a> to [[#invoke-observers]] with a list containing the <a>context object</a>
```
Comment (clelland) [inline index.src.html]: I think you might want to specify either that list is a new list, or that it *just* contains that object (as opposed to any list which contains that object).

You could also define an algorithm to notify a single observer, and then redefine `#invoke-observers` to call that inside of it's loop, but that might be overengineering it.

File: index.src.html [hunk 13]
```diff
@@ -769,103 +782,123 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
...
+
   The <dfn method for=ReportingObserver><code>disconnect()</code></dfn> method,
   when invoked, must run these steps:
-  
-  1. If present, remove the <a>context object</a> from the list of
-     <a>registered reporting observers</a> for the <a>environment settings
-     object</a> associated with the <a>context object</a>.
- 
+
+  1. If the <a>context object</a> is not <a>registered</a>, stop.
```
Comment (clelland) [inline index.src.html]: stop => return

File: index.src.html [hunk 14]
```diff
@@ -769,103 +782,123 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
...
-  
-  2. If it is not already present, add the <a>context object</a> to the 
-     list of <a>registered reporting observers</a> for the <a>environment
-     settings object</a> associated with the <a>context object</a>.
-  
+
+  2. Let |settings| be the <a>environment settings object</a> associated with
+     the <a>context object</a>
+
+  3. If it is not already registered, add the <a>context object</a> to
```
Comment (clelland) [inline index.src.html]: If you make the registered reporting observer list an ordered set, then you can rewrite this to say
"Append the context object to the registered reporting observer list of its associated environment settings object"

File: index.src.html [hunk 15]
```diff
@@ -769,103 +782,123 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
...
+  1. If the <a>context object</a> is not <a>registered</a>, stop.
+
+  2. Remove the <a>context object</a> from the <a>registered reporting observer
+     list</a> of its associated <a>environment settings object</a>.
+
+  <h3 id="add-report" algorithm>
+    Add |report| to reporting |observer|
+  </h3>
+
+  This algorithm adds |report| as a {{Report}} object to |observer|'s <a>report
```
Comment (clelland) [inline index.src.html]: Can you start this with something like "Given a report |report|, ..." (and link the first term) to make clear what "Report" type that argument is?

File: index.src.html [hunk 16]
```diff
@@ -769,103 +782,123 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
...
-     object</a> associated with the <a>context object</a>.
- 
+
+  1. If the <a>context object</a> is not <a>registered</a>, stop.
+
+  2. Remove the <a>context object</a> from the <a>registered reporting observer
+     list</a> of its associated <a>environment settings object</a>.
+
+  <h3 id="add-report" algorithm>
+    Add |report| to reporting |observer|
```
Comment (clelland) [inline index.src.html]: It seems odd to use the term "reporting observer" here, which is a type, but then turn just the second word into a parameter name. Probably just call this algorithm "Add |report| to |observer|", and mention below that |observer| is a ReportingObserver

File: index.src.html [hunk 17]
```diff
@@ -769,103 +782,123 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
...
+  1. If |observer|'s <a>options</a> has a {{ReportingObserverOptions/types}}
+     member which does not contain |report|'s [=report/type=], stop.
+
+  2. Create a new {{Report}} |r| with {{Report/type}} initialized to |report|'s
+     [=report/type=], {{Report/url}} initialized to |report|'s [=report/url=],
+     and {{Report/body}} initialized to |report|'s [=report/body=].
+
+  Issue: how to polymorphically initialize body?
+
+  3. Add |r| to the end of |observer|'s <a>report list</a>.
```
Comment (clelland) [inline index.src.html]: The infra standard for list defines 'append' as a list operation, which I think you can use here.

File: index.src.html [hunk 18]
```diff
@@ -769,103 +782,123 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
...
+
+  <h3 id="add-report" algorithm>
+    Add |report| to reporting |observer|
+  </h3>
+
+  This algorithm adds |report| as a {{Report}} object to |observer|'s <a>report
+  list</a>, so long as |report|'s <a>type</a> is observable by |observer|.
+
+  1. If |observer|'s <a>options</a> has a {{ReportingObserverOptions/types}}
+     member which does not contain |report|'s [=report/type=], stop.
```
Comment (clelland) [inline index.src.html]: stop => return

File: index.src.html [hunk 19]
```diff
@@ -769,103 +782,123 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
...
-     3. Add |r| to the end of |observer|'s <a>report list</a>.
-         
-  2. <a>Queue a task</a> to [[#invoke-observers]] with a copy of the
-     <a>registered observers</a> associated with |settings|.
+  1. For each {{ReportingObserver}} |observer| <a>registered</a> with
+     |settings|, execute [[#add-report]] on |report| and |observer|.
+ 
+  2. Add |report| to the <a>report buffer</a> associated with |settings|.
+  
+  3. If the <a>report buffer</a> now contains more than 100 reports, remove the
```
Comment (clelland) [inline index.src.html]: "The oldest report" implies that we're keeping timestamps or some other accounting information -- can you say something like 'remove the item at the beginning of report buffer'? (Since I think they're guaranteed to be stored in chronological order)

**Review by clelland — COMMENTED** [review]
Added a bunch of technical comments about the spec text. I'm still thinking about the way the algorithms run, and I'll have some more comments around that soon.

**Comment by clelland** [comment]
I think there's an odd interplay between the two levels of buffer going on --

If a number of reports (r1, r2, ..., rn) are present in the environment's report buffer, and then you run

    var ob = new ReportingObserver();
    ob.observe();
    ob.observe();


or even

    ob.observe();
    ob.disconnect();
    ...
    ob.observe(); // before the reports have been sent

then all of the reports in the buffer will be added to the observer's report list twice, and then the callback will be fired twice -- first with [r1, r2, ..., rn, r1, r2, ..., rn] as the *reports* list, and then with [] as the list. Is this intended?

* Is it intentional that futureOnly is false by default? (It feels like true is the more likely case) -- and should this be part of the 'options' dictionary that can change every time the observer is registered?
* Is it intentional that re-registering an observer can cause already-sent reports to be queued again?
* It seems like a bug, but should a single call to *callback* be able to contain multiple copies of a report, or be called with an empty report list?

File: index.src.html (hunk 9 above)
Comment (paulmeyer90) [reply index.src.html]: Yes. Actually, I would like to have the options set in the constructor rather than in observe(), they will now be initialized to the passed in parameter.

File: index.src.html (hunk 8 above)
Comment (paulmeyer90) [reply index.src.html]: Yes, that is what it should be.

File: index.src.html (hunk 14 above)
Comment (paulmeyer90) [reply index.src.html]: Done.

File: index.src.html (hunk 10 above)
Comment (paulmeyer90) [reply index.src.html]: Done.

File: index.src.html (hunk 11 above)
Comment (paulmeyer90) [reply index.src.html]: Done.

File: index.src.html (hunk 12 above)
Comment (paulmeyer90) [reply index.src.html]: I am going to move where these tasks are queued into the add-report algorithm, where that detail won't be necessary.

File: index.src.html (hunk 13 above)
Comment (paulmeyer90) [reply index.src.html]: Done.

File: index.src.html (hunk 16 above)
Comment (paulmeyer90) [reply index.src.html]: Done.

File: index.src.html (hunk 15 above)
Comment (paulmeyer90) [reply index.src.html]: Done.

File: index.src.html (hunk 18 above)
Comment (paulmeyer90) [reply index.src.html]: Done.

File: index.src.html (hunk 17 above)
Comment (paulmeyer90) [reply index.src.html]: Done.

File: index.src.html (hunk 19 above)
Comment (paulmeyer90) [reply index.src.html]: Okay, yeah.

File: index.src.html [hunk 20]
```diff
@@ -819,93 +819,94 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
...
-  the context of their own code.
-  
+
   1. For each {{ReportingObserver}} |observer| in |notify list|:
-  
+
      1. Let |reports| be a copy of |observer|'s <a>report list</a>
-     
+
      2. Clear |observer|'s <a>report list</a>
```
Comment (clelland) [inline index.src.html]: Clear => Empty

File: index.src.html [hunk 21]
```diff
@@ -819,93 +819,94 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
...
   This algorithm makes |report|'s contents available to any <a>registered</a>
   <a>reporting observers</a> on the provided <a>environment settings object</a>
   |settings|.
-  
+
   1. For each {{ReportingObserver}} |observer| <a>registered</a> with
      |settings|, execute [[#add-report]] on |report| and |observer|.
- 
+
   2. Add |report| to the <a>report buffer</a> associated with |settings|.
```
Comment (clelland) [inline index.src.html]: Add => Append

File: index.src.html [hunk 22]
```diff
@@ -819,93 +819,94 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
...
+
      1. Let |reports| be a copy of |observer|'s <a>report list</a>
-     
+
      2. Clear |observer|'s <a>report list</a>
-     
+
      3. <a spec=webidl>Invoke</a> |observer|'s <a>callback</a> with a list of
         arguments consisting of |reports| and |observer|, and |observer| as the
         <a>callback this value</a>.  If this throws an exception, <a>report the
```
Comment (clelland) [inline index.src.html]: Is it normal to include the observer as both an argument and the 'this' value?

File: index.src.html [hunk 23]
```diff
@@ -819,93 +819,94 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
...
-  3. Add |r| to the end of |observer|'s <a>report list</a>.
-      
+  3. Append |r| to |observer|'s <a>report list</a>.
+
+  4. If the size of |observer|'s <a>report list</a> is 1, <a>Queue a task</a> to
+     [[#invoke-observers]] with a copy of the <a>registered reporting observer
+     list</a> of the <a>environment settings object</a> associated with
+     |observer|.
+
   <h3 id="notify-observers" algorithm>
```
Comment (clelland) [inline index.src.html]: This algorithm feels like it's in the wrong spot... right now 5.1 calls 5.2, which calls 5.4. Then 4.1 calls this one (5.3) which also calls 5.2. I'm not sure where it should go, but maybe switching with `#add-report` would make sense? Then the calls would always flow 'down' in the document.

File: index.src.html [hunk 24]
```diff
@@ -819,93 +819,94 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
...
   observed behavior.
-  
-  Note: It is possible for the <a>registered</a> reporting observers and
-  their <a>options</a> to be different at this time than when reports were
-  generated. It is important that we rely only on the state at the time of
-  generation, for example to permit libraries to observe behavior only within
-  the context of their own code.
-  
+
   1. For each {{ReportingObserver}} |observer| in |notify list|:
```
Comment (clelland) [inline index.src.html]: I think we want to add something here that checks to see if the report list is empty, or else there are still conditions where `callback` will be called with an empty list (Because 5.2 -- which is called every time you `observe()` with a new not-future-only observer -- will queue a task to call this algorithm with a list of all registered observers, the second time this is called, some of the report lists will have been cleared)

**Review by clelland — COMMENTED** [review]
This is looking good -- a couple of nits, and one algorithm change to try to stop calling callbacks without any actual reports.

File: index.src.html (hunk 22 above)
Comment (paulmeyer90) [reply index.src.html]: This seems to be the same as is done with PerformanceObserverCallback.

https://w3c.github.io/performance-timeline/#dom-performanceobservercallback

File: index.src.html (hunk 23 above)
Comment (paulmeyer90) [reply index.src.html]: That makes sense. I've swapped them.

File: index.src.html (hunk 21 above)
Comment (paulmeyer90) [reply index.src.html]: Done.

File: index.src.html (hunk 24 above)
Comment (paulmeyer90) [reply index.src.html]: Okay, I've added a line to continue on empty list.

File: index.src.html (hunk 20 above)
Comment (paulmeyer90) [reply index.src.html]: Done.

**Comment by clelland** [comment]
This looks good -- thanks, @paulmeyer90
